ddtrace/contrib/langchain/patch.py

```diff
@@ -0,0 +1,713 @@
+from ddtrace.contrib.langchain.constants import text_embedding_models
+from ddtrace.contrib.langchain.constants import vectorstores
+from ddtrace.contrib.trace_utils import unwrap
+from ddtrace.contrib.trace_utils import with_traced_module
+from ddtrace.contrib.trace_utils import wrap
+from ddtrace.internal.agent import get_stats_url
+from ddtrace.internal.logger import get_logger
+from ddtrace.internal.utils.formats import asbool
+from ddtrace.pin import Pin
+
+
+if TYPE_CHECKING:
+    from ddtrace import Span
+
+
+log = get_logger(__name__)
+
+
+config._add(
+    "langchain",
+    {
+        "logs_enabled": asbool(os.getenv("DD_LANGCHAIN_LOGS_ENABLED", False)),
+        "metrics_enabled": asbool(os.getenv("DD_LANGCHAIN_METRICS_ENABLED", True)),
+        "span_prompt_completion_sample_rate": float(os.getenv("DD_LANGCHAIN_SPAN_PROMPT_COMPLETION_SAMPLE_RATE", 1.0)),
+        "log_prompt_completion_sample_rate": float(os.getenv("DD_LANGCHAIN_LOG_PROMPT_COMPLETION_SAMPLE_RATE", 0.1)),
+        "span_char_limit": int(os.getenv("DD_LANGCHAIN_SPAN_CHAR_LIMIT", 128)),
+        "_api_key": os.getenv("DD_API_KEY"),
+    },
+)
+
+
+class _LangChainIntegration(BaseLLMIntegration):
+    _integration_name = "langchain"
+
+    def __init__(self, config, stats_url, site, api_key):
+        super().__init__(config, stats_url, site, api_key)
+
+    def _set_base_span_tags(self, span):
+        # type: (Span) -> None
+        return None
+
+    def _logs_tags(self, span):
+        # type: (Span) -> str
+        api_key = span.get_tag("langchain.request.api_key") or ""
+        tags = "env:%s,version:%s,langchain.request.provider:%s,langchain.request.model:%s,langchain.request.type:%s,langchain.request.api_key:%s" % (  # noqa: E501
+            (config.env or ""),
+            (config.version or ""),
+            (span.get_tag("langchain.request.provider") or ""),
+            (span.get_tag("langchain.request.model") or ""),
+            (span.get_tag("langchain.request.type") or ""),
+            api_key,
+        )
+        return tags
+
+    def _metrics_tags(self, span):
+        provider = span.get_tag("langchain.request.provider") or ""
+        api_key = span.get_tag("langchain.request.api_key") or ""
+        tags = [
+            "version:%s" % (config.version or ""),
+            "env:%s" % (config.env or ""),
+            "service:%s" % (span.service or ""),
+            "langchain.request.provider:%s" % provider,
+            "langchain.request.model:%s" % (span.get_tag("langchain.request.model") or ""),
+            "langchain.request.type:%s" % (span.get_tag("langchain.request.type") or ""),
+            "langchain.request.api_key:%s" % api_key,
+            "error:%d" % span.error,
+        ]
+        err_type = span.get_tag("error.type")
+        if err_type:
+            tags.append("error_type:%s" % err_type)
+        return tags
+
+    def record_usage(self, span, usage):
+        if not usage or not self._config.metrics_enabled:
```

Comment:
Very minor nit: `ddtrace.config.somethingthatdoes_not_exist` will return a truthy value (it will initialize or refer to an IntegrationConfig object). It might be safer here to check if self._config.metrics_enabled is False.